### PR TITLE
Replace terms with property values that will cause Invalid Number Exceptions in Solr 6 with dummy terms in FtsNodeVisitor

### LINKED ISSUE #42: Search: Property values that can never match should be filtered out
It is currently possible to search for a property, with a value that is impossible to exist, due to type definitions in Alfresco. An easy example which can be used to reproduce this, is the `sys:node-dbid` property, which is a property of type `d:long`. Searching for that property with a value of `Long.MAX_VALUE + 1` will return an error. Searching with `Long.MAX_VALUE` succeeds and just returns an empty set (as no one has that many nodes).

```curl -v -u admin:admin localhost:32787/alfresco/s/apix/v1/search -XPOST -H 'Content-Type: application/json' -d '{"query":{"and":[{"or":[{"property":{"name":"sys:node-dbid","value":"9223372036854775808"}}]}]}}'```

Systems that query Alfresco (through Alfred Api for example) should not need to know about the datatype of properties that are in the custom models of the Alfresco. In our case, a system (Alfred Finder) expands a value to all sets of property name : that value
As a necessary workaround, Alfred Api can check the datatype using the Alfresco `DictionaryService` and filter out the terms that do not comply with the datatype of their property value, before sending out the query to Alfresco.

### PR DESCRIPTION
**Pull request by driesdeproost**
Fixes https://xenitsupport.jira.com/browse/ALFREDAPI-427
Fixes #42 

- [x] Is [CHANGELOG.md](https://github.com/xenit-eu/alfred-api/blob/master/CHANGELOG.md) extended?
- [x] Does this PR avoid breaking the API? 
    Breaking changes include adding, changing or removing endpoints and/or JSON objects used in requests and responses.
- [x] Does the PR comply to REST HTTP result codes policy outlined in the [developer guide](https://github.com/xenit-eu/alfred-api/blob/master/developer-documentation)?
- [x] Does the PR follow our [coding styleguide and other active procedures](https://xenitsupport.jira.com/wiki/spaces/XEN/pages/624558081/XeniT+Enhancement+Proposals+XEP)?
- [x] Is usage of `this.` prefix avoided?

See [README.md](./README.md) for full explanation.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,8 @@
 ### Fixed
 * [ALFREDAPI-419](https://xenitsupport.jira.com/browse/ALFREDAPI-419): Add urldecoding to deleteAssociation endpoint
 * [ALFREDAPI-422](https://xenitsupport.jira.com/browse/ALFREDAPI-422): Fix totalResults and limits for TDMQ's
+* [ALFREDAPI-427](https://xenitsupport.jira.com/browse/ALFREDAPI-427): Replace terms with property values that will cause Invalid Number Exceptions in Solr 6 with dummy terms in FtsNodeVisitor
```
Comment (tgeens) [inline CHANGELOG.md:16]: @driesdeproost the changelog items should state what issue this solves (audience:user), not how (audience: developer)

File: apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java [hunk 2]
```diff
@@ -322,4 +321,31 @@ private void withLocale(Locale locale, AuthenticationUtil.RunAsWork<Object> work
...
+    public void testNotFilteredOut() {
+        SearchSyntaxNode node =
+                new QueryBuilder()
+                        .startOr()
+                        .property("{http://www.alfresco.org/model/system/1.0}node-dbid", "9223372036854775807")
+                        .end()
+                        .create();
+        SearchQuery searchQuery = new SearchQuery();
+        searchQuery.setQuery(node);
+        searchService.query(searchQuery);
```
Comment (tgeens) [inline apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:335]: what is being tested here ? if it is "does not throw exception", either:
- put that in the method name
- assert that the return value is not null

File: apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java [hunk 3]
```diff
@@ -322,4 +321,31 @@ private void withLocale(Locale locale, AuthenticationUtil.RunAsWork<Object> work
...
+    }
+
+    @Test
+    public void testNothingLeftAfterFiltering() {
```
Comment (tgeens) [inline apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:339]: this is EXACTLY the same test as above ?

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            logger.warn("Instantiating PropertyTypeCheckService without a PropertyService, type checks will always succeed");
+        }
+        this.propertyService = propertyService;
+    }
+
+    public boolean fitsType(PropertySearchNode node) {
+        if (node == null || node.getValue() == null || node.getName() == null || propertyService == null) {
+            return true;
+        }
+        QName qName = new QName(node.getName().replaceAll("\\\\", "")); //stuff like hyphen in node-dbid is escaped
```
Comment (tgeens) [inline apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:37]: isn't there an API-way to do this ? feels like we're missing potential edge cases here ?

File: apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java (hunk 2 above)
Comment (driesdeproost) [reply apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:335]: exactly, does not throw exception
agreed

File: apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java (hunk 3 above)
Comment (driesdeproost) [reply apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:339]: It might seem so, but the long value is 1 greater, crossing the max value boundary

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.Map.Entry;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PropertyTypeCheckService {
+
+    private final static Logger logger = LoggerFactory.getLogger(PropertyTypeCheckService.class);
+
+    private final Map<String, Function<String, Boolean>> constraintMap = new HashMap<String, Function<String, Boolean>>() {{
```
Comment (tgeens) [inline apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:17]: constraint is an overloaded term here
Suggestions: `dataTypePredicates`

Comment (tgeens) [inline apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:17]: Replace `Function<String, Boolean>` with `Predicate<String>` ?

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java (hunk 4 above)
Comment (driesdeproost) [reply apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:37]: Good question, I will look into it

File: apix-impl/src/test/java/eu/xenit/apix/tests/search/FtsNodeVisitorTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            querySyntaxTree = querySyntaxTree.property(property, value);
+        }
+        return querySyntaxTree
+                .end()
+                .term("type", "{http://www.alfresco.org/model/content/1.0}content")
+                .end()
+                .create();
+    }
+
+    private PropertyService generatePropertyServiceMock() {
```
Comment (tgeens) [inline apix-impl/src/test/java/eu/xenit/apix/tests/search/FtsNodeVisitorTest.java:76]: suggestion: create a `PropertyServiceStub` (inner?) class instead that accepts that `propertyToDataType` map as constructor arg

File: apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java (hunk 3 above)
Comment (tgeens) [reply apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:339]: Ok, need to make that more obvious somehow ;-)

Comment (driesdeproost) [reply apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:339]: I will add it to the method name, together with the `throws exception`

File: apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java [hunk 7]
```diff
@@ -322,4 +321,31 @@ private void withLocale(Locale locale, AuthenticationUtil.RunAsWork<Object> work
      */
     abstract protected void withTestFacets(AuthenticationUtil.RunAsWork<Object> work) throws Exception;
 
+
+    @Test
+    public void testNotFilteredOut() {
+        SearchSyntaxNode node =
+                new QueryBuilder()
+                        .startOr()
+                        .property("{http://www.alfresco.org/model/system/1.0}node-dbid", "9223372036854775807")
```
Comment (kpiot123) [inline apix-impl/src/integration-test/java/eu/xenit/apix/tests/search/SearchServiceTest.java:330]: Can you use Long.MAX_VALUE instead of "9223372036854775807" so people know what the value means?

File: apix-impl/src/test/java/eu/xenit/apix/tests/search/FtsNodeVisitorTest.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        PropertyService propertyService = mock(PropertyService.class);
```
Comment (kpiot123) [inline apix-impl/src/test/java/eu/xenit/apix/tests/search/FtsNodeVisitorTest.java:77]: @driesdeproost PropertyService is an service of Alfred API so you don't need to mock it. It is better to create a stub like Toon said and then mock the Alfresco services you need. In this case you will need a mock for the dictionary service I believe.

File: CHANGELOG.md (hunk 1 above)
Comment (driesdeproost) [reply CHANGELOG.md]: done

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    public boolean fitsType(PropertySearchNode node) {
+        if (node == null || node.getValue() == null || node.getName() == null || propertyService == null) {
+            return true;
+        }
+        QName qName = new QName(node.getName().replaceAll("\\\\", "")); //stuff like hyphen in node-dbid is escaped
+        PropertyDefinition propertyDefinition = propertyService.GetPropertyDefinition(qName);
+        if (propertyDefinition == null || propertyDefinition.getDataType() == null) {
+            return true;
+        }
+        for (Entry<String, Predicate<String>> entry : dataTypePredicates.entrySet()) {
```
Comment (RVanhuysseXenit) [inline apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:43]: Why do you iterate over the entrySet and compare keyvalues, instead of just calling `dataTypePredicates.get()`?

Comment (driesdeproost) [reply apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:43]: That's better :+1:

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java (hunk 4 above)
Comment (driesdeproost) [reply apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java]: Didn't find anything with some quick searches
I abstracted out the existing escape mechanism and added an inverse (and some tests on them) as a proposal
https://github.com/xenit-eu/alfred-api/pull/44/commits/2aac5d36004c0c3b1c02f51e9a571f451931d321

File: CHANGELOG.md (hunk 1 above)
Comment (tgeens) [reply CHANGELOG.md]: Imo we should drop the "in FtsNodeVisitor", but maybe @kpiot123 or @WimCrols can give their opionion

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java (hunk 4 above)
Comment (tgeens) [reply apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java]: See comment on that commit.

File: CHANGELOG.md (hunk 1 above)
Comment (kpiot123) [reply CHANGELOG.md]: Yes, you can drop that part. People reading this don't know what an FtsNodeVisitor is.

File: apix-impl/src/test/java/eu/xenit/apix/tests/search/FtsNodeVisitorTest.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
+package eu.xenit.apix.tests.search;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.Mockito.mock;
```
Comment (kpiot123) [inline apix-impl/src/test/java/eu/xenit/apix/tests/search/FtsNodeVisitorTest.java:6]: @driesdeproost You don't seem to be using mocks anymore so you can remove this import

File: CHANGELOG.md (hunk 1 above)
Comment (driesdeproost) [reply CHANGELOG.md]: Actually, what you see here is the outdated
Have a look at https://github.com/xenit-eu/alfred-api/pull/44/files

File: apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Map;
+import java.util.function.Predicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PropertyTypeCheckService {
+
+    private final static Logger logger = LoggerFactory.getLogger(PropertyTypeCheckService.class);
+
+    private final Map<String, Predicate<String>> dataTypePredicates = new HashMap<String, Predicate<String>>() {{
```
[the remark this answers is not in the document]
Comment (driesdeproost) [reply apix-impl/src/main/java/eu/xenit/apix/alfresco/dictionary/PropertyTypeCheckService.java:16]: Thanks, nice catch
